### PR TITLE
Isa fixes

### PR DESCRIPTION
**Pull request by donofrio**
Fixes for ISA instructions as well as the instruction pre-fetcher

### CODE REVIEW

File: include/insns/RV64I.h [hunk 1]
```diff
@@ -192,11 +192,11 @@ namespace SST{
...
+      {RevInstEntryBuilder<RevInstDefaults>().SetMnemonic("sllw %rd, %rs1, %rs2" ).SetCost(1).SetOpcode( 0b0111011).SetFunct3(0b001).SetFunct7(0b0000000).SetrdClass(RegGPR).Setrs1Class(RegGPR).Setrs2Class(RegGPR    ).Setrs3Class(RegUNKNOWN).Setimm(FUnk).SetFormat(RVTypeR).SetImplFunc(&sllw ).InstEntry},
+      {RevInstEntryBuilder<RevInstDefaults>().SetMnemonic("srlw %rd, %rs1, %rs2" ).SetCost(1).SetOpcode( 0b0111011).SetFunct3(0b101).SetFunct7(0b0000000).SetrdClass(RegGPR).Setrs1Class(RegGPR).Setrs2Class(RegGPR    ).Setrs3Class(RegUNKNOWN).Setimm(FUnk).SetFormat(RVTypeR).SetImplFunc(&srlw ).InstEntry},
+      {RevInstEntryBuilder<RevInstDefaults>().SetMnemonic("sraw %rd, %rs1, %rs2" ).SetCost(1).SetOpcode( 0b0111011).SetFunct3(0b101).SetFunct7(0b0100000).SetrdClass(RegGPR).Setrs1Class(RegGPR).Setrs2Class(RegGPR    ).Setrs3Class(RegUNKNOWN).Setimm(FUnk).SetFormat(RVTypeR).SetImplFunc(&sraw ).InstEntry}
       };
```
Comment (donofrio) [inline include/insns/RV64I.h:202]: these are all 2-operand functions and had their rs2 class set to UNKNOWN which makes the decoder ignore them. I updated to be RegGPR

File: src/RevPrefetcher.cc [hunk 2]
```diff
@@ -156,6 +156,7 @@ void RevPrefetcher::Fill(uint64_t Addr){
   if((Addr%4)!=0){
     // not 32bit aligned, adjust the base address by 2 bytes
     Fill(Addr&0xFFFFFFFFFFFFFFFC);
+    return;
   }
```
Comment (donofrio) [inline src/RevPrefetcher.cc:160]: w/o this exit we end up filling twice... once on the aligned address, and a 2nd time on the unaligned address

File: include/insns/RV64I.h [hunk 3]
```diff
@@ -144,7 +144,8 @@ namespace SST{
       }
 
       static bool addw(RevFeature *F, RevRegFile *R,RevMem *M,RevInst Inst) {
-        R->RV64[Inst.rd] = dt_u64(td_u64(R->RV64[Inst.rs1],64) + td_u64(R->RV64[Inst.rs2],64),64);
+        R->RV64[Inst.rd] = dt_u32(td_u32(R->RV64[Inst.rs1],32) + td_u32(R->RV64[Inst.rs2],32),32); //addw operates on and produces 32-bit results even on RV64I codes 
+        SEXTI(R->RV64[Inst.rd], 32);    //Sign extend the result up to 64bits
```
Comment (donofrio) [inline include/insns/RV64I.h:148]: addw operates explicitly on 32-bit values so the input casting was updated an then  we sign extend the result

File: include/insns/RV64I.h [hunk 4]
```diff
@@ -151,7 +151,8 @@ namespace SST{
       }
 
       static bool subw(RevFeature *F, RevRegFile *R,RevMem *M,RevInst Inst) {
-        R->RV64[Inst.rd] = dt_u64(td_u64(R->RV64[Inst.rs1],64) - td_u64(R->RV64[Inst.rs2],64),64);
+        R->RV64[Inst.rd] = dt_u32(td_u32(R->RV64[Inst.rs1],32) - td_u32(R->RV64[Inst.rs2],32),32);
+        SEXTI(R->RV64[Inst.rd], 32);    //Sign extend the result up to 64bits
         R->RV64_PC += Inst.instSize;
```
Comment (donofrio) [inline include/insns/RV64I.h:156]: subw operates on 32-bit values - updated inputs and sign extend output.

File: src/RevProc.cc [hunk 5]
```diff
@@ -1408,7 +1408,7 @@ RevInst RevProc::DecodeInst(){
   // Stage 4: Determine if we have a funct7 field (R-Type)
   uint32_t Funct7 = 0x00ul;
   if( inst65 == 0b01 ) {
-    if( (inst42 == 0b011) || (inst42 == 0b100) ){
+    if( (inst42 == 0b011) || (inst42 == 0b100) || (inst42 == 0b110) ){
       // R-Type encodings
```
Comment (donofrio) [inline src/RevProc.cc:1416]: without this additional check for funct7 subw was decoding as addw

File: include/insns/RV64I.h [hunk 6]
```diff
@@ -193,7 +193,7 @@ namespace SST{
...
-      {RevInstEntryBuilder<RevInstDefaults>().SetMnemonic("sraiw %rd, %rs1, $imm").SetCost(1).SetOpcode( 0b0011011).SetFunct3(0b101).SetFunct7(0b0100000).SetrdClass(RegGPR).Setrs1Class(RegGPR).Setrs2Class(RegGPR    ).Setrs3Class(RegUNKNOWN).Setimm(FUnk).SetFormat(RVTypeR).SetImplFunc(&sraiw ).InstEntry},
+      {RevInstEntryBuilder<RevInstDefaults>().SetMnemonic("sraiw %rd, %rs1, $imm").SetCost(1).SetOpcode( 0b0011011).SetFunct3(0b101).SetFunct7(0b0100000).SetrdClass(RegGPR).Setrs1Class(RegGPR).Setrs2Class(RegUNKNOWN).Setrs3Class(RegUNKNOWN).Setimm(FImm).SetFormat(RVTypeR).SetImplFunc(&sraiw ).InstEntry},
       {RevInstEntryBuilder<RevInstDefaults>().SetMnemonic("addw %rd, %rs1, %rs2" ).SetCost(1).SetOpcode( 0b0111011).SetFunct3(0b000).SetFunct7(0b0000000).SetrdClass(RegGPR).Setrs1Class(RegGPR).Setrs2Class(RegGPR    ).Setrs3Class(RegUNKNOWN).Setimm(FUnk).SetFormat(RVTypeR).SetImplFunc(&addw ).InstEntry},
```
Comment (donofrio) [inline include/insns/RV64I.h:197]: sraiw is a special case R-type with a 6 bit immd field. I updated the table for use in the decoder. Note that I need to likely fix srliw and slliw, just haven't gotten to those asm tests yet

File: src/RevProc.cc [hunk 7]
```diff
@@ -995,7 +995,11 @@ RevInst RevProc::DecodeRInst(uint32_t Inst, unsigned Entry){
   }
 
   // imm
-  DInst.imm     = 0x0;
+  if( (InstTable[Entry].imm == FImm) && (InstTable[Entry].rs2Class == RegUNKNOWN)){
+    DInst.imm  = DECODE_IMM12(Inst) & 0b011111; 
+  }else{
+    DInst.imm     = 0x0;
+  }
```
Comment (donofrio) [inline src/RevProc.cc:1003]: Handle the special R-type that has the shift amount (this seems like the best place to do this?)

File: src/RevProc.cc [hunk 8]
```diff
@@ -1415,6 +1419,9 @@ RevInst RevProc::DecodeInst(){
   }else if((inst65== 0b10) && (inst42 == 0b100)){
       // R-Type encodings
       Funct7 = ((Inst >> 25) & 0b1111111);
+  }else if((inst65 == 0b00) && (inst42 == 0b110) && (Funct3 != 0)){
+      // R-Type encodings
+      Funct7 = ((Inst >> 25) & 0b1111111);
```
Comment (donofrio) [inline src/RevProc.cc:1424]: correct decoder to properly decode sraiw

**Review by jleidel — COMMENTED** [review]
looks solid

**Pull request merged**